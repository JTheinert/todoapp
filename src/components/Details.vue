<template>
  <div>

    <!-- THIS IS THE TODO ITEM -->
    <div v-show="todo.isOpen">
      <div class="todo-item-middle">
        <div
          v-if="!todo.editing"
          class="todo-item-lbel"
          @dblclick="editTodo(todo)"
        >

          <item-title/>
        </div>
        <!-- THIS TRIGGERS THE EDIT METHODS -->
        <input
          v-focus
          v-if="todo.editing"
          v-model="todo.title"
          class="todo-item-edit"
          type="text"
          @keyup.enter="doneEdit(todo)"
          @blur="doneEdit(todo)"
          @keyup.escape="cancelEdit(todo)"
        >

        <div v-show="todo.isOpen">
          <p class="todo-description-edit">Item NR. {{ todo.id }}</p>

          <!-- THIS IS THE INPUT FIELD FOR THE EDIT -->
          <input
            v-focus
            v-model="todo.description"
            class="todo-description-edit"
            type="text"
            placeholder="Edit Description"
            @keyup.enter="doneEditDescription(todo)"
            @blur="doneEditDescription(todo)"
            @keyup.escape="cancelEditDescription(todo)"
          >
          <todo-selection/>

          <!-- With Opened Detils -->
          <div
            v-show="todo.isOpen"
            class="details-item-right">
            <!-- THIS IS THE DELETE BUTTON IN DETILS-->

            <delete-button/>
            <close-button/>
            <!-- THIS IS THE CLOSE BUTTON IN DETILS-->

          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import statusItem from './Overview/statusItem'
import itemTitle from './Overview/itemTitle'
import deleteButton from './Overview/DeleteButton'
import closeButton from './Details/detailsClose'
import todoSelection from './Details/chooseStatus'

export default {
  name: 'Details',
  components: { itemTitle, todoSelection, statusItem, deleteButton, closeButton },

  directives: {
    focus: {
      inserted: function (el) {
        el.focus()
      }
    }
  },

  props: {

  },
  data () {
    return {
      todo: this.$store.state.todos[this.$route.params.id]
    }
  },
  computed: {
    todos () {
      return this.$store.state.todos
    },
    enbaleall () {
      return this.$store.state.enableall
    },
    filteredNr () {
      return this.$store.state.filteredNr
    }
  },
  methods: {

    editTodo (todo) {
      todo.editing = true
      this.beforeEditCache = todo.title
    },
    editDescription (todo) {
      todo.descriptionEditing = true
      this.beforeEditDescriptionCache = todo.description
    },
    doneEdit (todo) {
      if (todo.title.trim().length === 0) {
        todo.title = this.beforeEditCache
      }
      todo.editing = false
    },
    doneEditDescription (todo) {
      if (todo.description.trim().length === 0) {
        todo.description = this.beforeEditDescriptionCache
      }
      todo.descriptionEditing = false
    },
    cncelEdit (todo) {
      todo.title = this.beforeEditCache
      todo.editing = false
    },
    cncelEditDescription (todo) {
      todo.description = this.beforeEditDescriptionCache
      todo.descriptionEditing = false
    },

    cancelDelete () {
      this.enableall = !this.enableall
    },
    openModal (index, todo) {
      // eslint-disable-next-line no-unused-expressions
      this.$router.push('delete')
      this.$store.commit('updateNummer', {
        nummer: todo.id
      })
    },
    openModlDetails (index, todo) {
      this.$store.commit('updateIsOpen', {
        id: todo.id,
        isOpen: todo.isOpen
      })
      // eslint-disable-next-line no-unused-expressions
      // eslint-disable-next-line func-call-spacing
      this.$router.push('/delete')(todo.isOpen = !todo.isOpen)

      todo.buttonName = todo.isOpen ? 'Close' : 'Details'
      this.nummer = index
    }

  }
}
</script>

<style>
.todo-item {
margin-bottom: 3px;
display: inline-block;
align-items: center;
justify-content: space-between;
animation-duration: 0.3s;
color: #29cc;
border: 1px solid lavenderblush;
}

.details-item-right {
display: flex;
text-align: center;
margin-right: 10px;
margin-bottom: 5px;
justify-content: space-between;
}

.status-item {
display: block;
align-items: left;
border: 1px;
font-size: 20px;
text-align: center;
}
.todo-item-label {
padding: 10px;
border: 1px solid white;
margin-left: 0px;
font-size: 35px;
font-style: italic;
font-weight: 800;
text-align: center;
}

.todo-item-edit {
font-size: 35px;
color: #29cc;
margin-left: auto;
width: 100%;
padding: 10px;
border: 1px solid #ccc;
font-family: "Rockwell Extr Bold", "Rockwell Bold", monospace;
text-align: center;
justify-content: center;
font-weight: 800;
}
.todo-description-edit {
font-size: 20px;
font-style: italic;
font-family: "Rockwell Extr Bold", "Rockwell Bold", monospace;
color: #29cc;
margin-left: 5px;
width: 400px;
padding: 5px;
display: block;
align-items: center;
border-radius: 5px;
}
.details-delete-button {
background-color: red;
font-size: 20px;
border-radius: 10px;
text-align: center;
color: white;
margin-left: 10px;
}

.done {
color: green;
}
.progress {
color: darkgoldenrod;
}
</style>
