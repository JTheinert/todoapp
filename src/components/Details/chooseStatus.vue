<template>
  <div
    class="todo-selection">

    <!-- THIS IS THE SELECTION FOR IN PROGRESS -->
    <div class="todo-select">
      <input
        id="1"
        v-model="todo.used"
        type="radio"
        class="todo-selection-progress"
        value="1"
        name="inProgess"
        @click="setFilterProgress(todo)"
      >
      <label for="1">in progress</label>
    </div>
    <div class="todo-select">
      <input
        id="2"
        v-model="todo.used"
        type="radio"
        class="todo-selection-done"
        value="2"
        name="Done"
        @click="setFilterDone(todo)"
      >
      <label for="2">done</label>
    </div>
  </div>

</template>

<script>
export default{
  data () {
    return {
      todo: this.$store.state.todos[this.$route.params.id]
    }
  },
  computed: {
    todos () {
      return this.$store.state.todos
    },
    enableAll () {
      return this.$store.state.enableAll
    },
    filteredNr () {
      return this.$store.state.filteredNr
    }
  },
  methods: {
    setFilterProgress (todo) {
      todo.filterNr = 2
      todo.filter = true
    },
    setFilterDone (todo) {
      todo.filterNr = 3
      todo.filter = true
    }
  }
}

</script>
<style>
.todo-selection-done {
font-size: 20px;
color: #2a9cac;
margin-left: 10px;
padding: 5px;
margin-top: 10px;
margin-bottom: 10px;
font-family: "Rockwell Extra Bold", "Rockwell Bold", monospace;
align-items: center;
display: inline-grid;
}
.todo-selection-progress {
font-size: 20px;
color: #2a9cac;
margin-left: 10px;
padding: 5px;
margin-top: 10px;
margin-bottom: 10px;
font-family: "Rockwell Extra Bold", "Rockwell Bold", monospace;
align-items: center;
display: inline-grid;
}

.todo-select {
font-size: 20px;
color: #2a9cac;
margin-left: 10px;
width: 400px;
padding: 5px;
margin-top: 20px;
margin-bottom: 20px;
font-family: "Rockwell Extra Bold", "Rockwell Bold", monospace;
align-items: center;
display: inline-flex;
}
</style>
