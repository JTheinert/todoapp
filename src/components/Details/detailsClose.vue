<template>
  <div>
    <button
      :id="index"
      class="close-details-button"
      name="closeButton"
      @click="closeDetails(todo)"
    >
      Close
    </button>
  </div>
</template>

<script>
export default {
  props: {
    index: {
      type: Number,
      default: 0
    }
  },
  data () {
    return {
      todo: this.$store.state.todos[this.$route.params.id]
    }
  },
  computed: {
    todoList () {
      return this.$store.state.todos
    }

  },
  methods: {
    closeDetails (todo) {
      todo.isOpen = false
      this.$store.commit('updateIsOpen', {
        id: todo.id,
        isOpen: todo.isOpen
      })
      this.$router.push({ path: '/' })
    }
  }
}
</script>
<style>
.close-details-button {
background-color: #2a9cac;
font-size: 20px;
border-radius: 10px;
text-align: center;
color: white;
margin-left: 12px;
}
</style>
