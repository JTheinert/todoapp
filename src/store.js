/* eslint-disable no-console */
import Vue from 'vue'
import Vuex from 'vuex'

Vue.use(Vuex)

export const store = new Vuex.Store({
  mutations: {
    updateIsOpen (state, { id, isOpen }) {
      const todo = state.todos.find(obj => obj.id === id)
      todo.isOpen = isOpen
    },
    updateFilteredNr () {
      // eslint-disable-next-line no-use-before-define
      const filteredNr = filteredNr
    },

    set (newTodo) {
      return newTodo
    },
    updateNummer (state, { nummer }) {
      state.nummer = nummer
    },
    updateIdForTodo (state) {
      state.idForTodo++
    }
  },
  methods: {
    mounted () {
      // eslint-disable-next-line no-console
      console.log('App mounted!')
      if (localStorage.getItem('todos')) { this.todos = JSON.parse(localStorage.getItem('todos')) }
    },
    watch: {
      todos: {
        handler () {
          console.log('Todos changed!')
          localStorage.setItem('todos', JSON.stringify(this.todos))
        },
        deep: true
      }
    }
  },
  state: {
    filteredNr: 1,
    modalShow: false,
    newTodo: '',
    idForTodo: 1,
    beforeEditCache: '',
    beforeEditDescriptionCache: '',
    placeholder: 'Edit Description',
    dialog: false,
    nummer: 0,
    todos: [
      {
        id: 0,
        title: 'cool',
        editing: false,
        descriptionEditing: false,
        isOpen: false,
        status: "isn't done",
        description: '',
        filter: false,
        filterNr: 1,
        used: 3
      }
    ]
  },

  getters: {
    doneTodos: state => {
      return state.todos.filter(todo => todo.done)
    },
    doneTodosCount: (getters) => {
      return getters.doneTodos.length
    },
    getTodoById: state => id => {
      return state.todos.find(todo => todo.id === id)
    },
    getName: () => 'rajesh'
  }
})
