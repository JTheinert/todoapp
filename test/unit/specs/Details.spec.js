import Details from '@/src/Components/Details.vue'

import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import { store } from '@/store'
import VueRouter from 'vue-router'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Details.vue', () => {
  test('Test', () => {
    expect(true).toBe(true)
  })
  const wrapper = mount(Details, {store, localVue, VueRouter})

  test('Title ?', () => {
    expect(wrapper.text()).toContain('cool')
  })

  test('Close button', () => {
    expect(wrapper.text()).toContain('Close')
  })

  test('Delete button', () => {
    expect(wrapper.text()).toContain('Delete')
  })

  wrapper.vm.$nextTick()

  test('Close Details function', () => {
    const closeDetails = jest.fn()

    wrapper.setMethods({
      closeDetails: closeDetails
    })

    wrapper.find('button.close-details-button').trigger('click')

    expect(closeDetails).toHaveBeenCalled()
  })

  test('open Details modal', () => {
    const openModalDetails = jest.fn()

    wrapper.setMethods({
      openModalDetails: openModalDetails
    })

    wrapper.find('button.details-delete-button').trigger('click')

    expect(openModalDetails).toHaveBeenCalled()
  })

  test('set Filter to Progress', () => {
    const setFilterProgress = jest.fn()

    wrapper.setMethods({
      setFilterProgress: setFilterProgress
    })

    wrapper.find('input.todo-selection-progress').trigger('click')

    expect(setFilterProgress).toHaveBeenCalled()
  })

  test('set Filter to Done', () => {
    const setFilterDone = jest.fn()

    wrapper.setMethods({
      setFilterDone: setFilterDone
    })

    wrapper.find('input.todo-selection-done').trigger('click')

    expect(setFilterDone).toHaveBeenCalled()
  })
  test('renders correctly', () => {
    const wrapper = mount(Details, {store})
    expect(wrapper.element).toMatchSnapshot()
  })
})
