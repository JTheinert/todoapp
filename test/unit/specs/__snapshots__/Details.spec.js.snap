// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details.vue renders correctly 1`] = `
<div>
  <div
    class="todo-item"
    id="0"
  >
    <div
      style="display: none;"
    >
      <div
        class="todo-item-middle"
      >
        <div
          class="todo-item-label"
        >
          <div
            class="todo-item-title"
          >
            cool
          </div>
           
          <!---->
           
          <div
            class="status-item"
          >
            <p
              class="done"
            >
              Done
            </p>
          </div>
        </div>
         
        <div
          style="display: none;"
        >
          <p
            class="todo-description-edit"
          >
            Item NR. 0
          </p>
           
          <input
            class="todo-description-edit"
            placeholder="Edit Description"
            type="text"
          />
           
          <div
            class="todo-selection"
          >
            <div
              class="todo-select"
            >
              <input
                class="todo-selection-progress"
                id="1"
                name="inProgess"
                type="radio"
                value="1"
              />
               
              <label
                for="1"
              >
                in progress
              </label>
            </div>
             
            <div
              class="todo-select"
            >
              <input
                class="todo-selection-done"
                id="2"
                name="Done"
                type="radio"
                value="2"
              />
               
              <label
                for="2"
              >
                done
              </label>
            </div>
          </div>
        </div>
         
        <div
          class="details-item-right"
          style="display: none;"
        >
          <button
            class="details-delete-button"
            id="0"
          >
            	
            Delete
          
          </button>
           
          <button
            class="close-details-button"
            id="0"
            name="closeButton"
          >
            
            Close
          
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
